### ISSUE TITLE
Use streamable buffers for the collaboration buffer as well

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#965]
Today the collaboration plugin requests all buffers on initial connection. For a project containing many GB of data this will both lock the "main thread" of the server for a considerable time since it will compress all these buffers (this could be optimized by threading the compression, but that is beside the point of this feature request) and mean that the client can't be used until all data is downloaded, which could also take a very long time if the network isn't top notch.

If the collaboration connection could create streamable buffers instead, then only the buffers which are actually used would be fetched, and the files could be compressed one by one and placed in the collaboration cache as compressed files, and the server would be able to tick between the network requests.

**Comment by simon-ourmachinery** [comment]
related or duplicate of #863

**Comment by niklas-ourmachinery** [comment]
Agreed. But this could potentially still cause big stalls, depending on the size of the initial buffer data set that has to be sent.

I think even better would be to make the code robust to handling "empty" buffers and fill them on demand. Kind of like how textures work right now. If a model needs a compressed texture that isn't available we will just not show anything until it has been generated.

If this could be extended to work with all buffers, then both the client and the server would be able to run without major stalls. The scene would just not show anything until the required buffers had been downloaded.

However, this means that all code that deals with buffer must be able to handle a zero-sized buffer and to update that buffer when the data is arrived.

**Comment by Tisten** [comment]
Yes, there would be stalls, how big would depend on how the assets are laid out in the project. In my own setup I already experience these stalls, and in my case they are to prefer over the initial connection to collaboration.

I'm storing my machinery project on Google drive, mounted via their Windows Desktop GoogleFS reparse point based virtual file system. Every time I access a new or updated file (not added by me) it will be downloaded from google before the data is handed to the application. This adds a stall on every non-cached file access in the same way as streaming buffers in collaboration would.

I would also like to see systems using big buffers accepting zero-sized buffers, but I don't think it needs to be black-or-white. To start with a system could flag if it can handle zero-buffers, and those which can't will stall instead.

One of the most memory consuming types of buffers in my project are of the "tm_graph_cached_node_result" and "tm_shader_system_compiled_result" types, and I imagine that the machinery can run withour these. It might even go faster to regenerate them than to download them from another machine if the network is slow.

And I think the number of those buffers comes partially from this bug: https://github.com/OurMachinery/themachinery-public/issues/943

**Comment by niklas-ourmachinery** [comment]
I agree that multiple smaller stalls are preferable to one big stall.

Good point about the black-and-whiteness... yes, incrementally transitioning to zero-sized buffers is preferable.

**Comment by Tisten** [comment]
@niklas-ourmachinery I tried the "tm_on_demand_buffer_streaming" feature today.

The first issue I got was that I wanted to be able to both use clients with that feature on, and off, against the same server. Since the "collaboration_state_id" hand shake string differs depending on the feature, this doesn't work. My hope is that to really battletest this feature, a select few should be able to turn the feature on while the rest runs as normal, and they all run against the same server.

Second feature I got directly after connecting to a local server was a deadlock on the client. Only one thread has a callstack, and it is stuck trying to lock an atomic flag `tm_creation_graph_node_cache_t::lock` here, but the flag is already set to one:
```
>	tm_creation_graph.dll!lock_resource_cache(tm_the_truth_o * tt, tm_tt_id_t graph_id, tm_tt_id_t node_id) Line 4106	C
 	tm_dcc_asset.dll!load_dcc_image_task(void * data, unsigned __int64 task_id) Line 684	C
 	tm_dcc_asset.dll!dcc_image_node_f(tm_creation_graph_interpreter_context_t * ctx) Line 774	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 431	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 424	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 424	C
 	tm_creation_graph.dll!trigger_event(tm_creation_graph_instance_t * rc, unsigned __int64 e) Line 461	C
 	tm_creation_graph.dll!trigger_graph_event(tm_creation_graph_instance_t * inst, tm_creation_graph_context_t * context, unsigned __int64 event) Line 2695	C
 	tm_creation_graph.dll!refresh_outputs(tm_creation_graph_instance_t * inst, tm_creation_graph_context_t * context) Line 5027	C
 	tm_creation_graph.dll!output(tm_creation_graph_instance_t * inst, unsigned __int64 node_name_hash, tm_creation_graph_context_t * context, const tm_creation_graph_output_node_type_t * * output_node_type) Line 5058	C
 	tm_creation_graph.dll!import_node_r(tm_creation_graph_interpreter_context_t * ctx) Line 1394	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 431	C
 	tm_creation_graph.dll!trigger_event(tm_creation_graph_instance_t * rc, unsigned __int64 e) Line 461	C
 	tm_creation_graph.dll!trigger_graph_event(tm_creation_graph_instance_t * inst, tm_creation_graph_context_t * context, unsigned __int64 event) Line 2695	C
 	tm_creation_graph.dll!trigger_event(tm_creation_graph_instance_t * inst, unsigned __int64 event_id, tm_creation_graph_context_t * context) Line 4803	C
 	tm_creation_graph.dll!create_instance(tm_the_truth_o * tt, tm_tt_id_t graph, tm_creation_graph_context_t * context) Line 4711	C
 	tm_creation_graph.dll!import_node_r(tm_creation_graph_interpreter_context_t * ctx) Line 1390	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 431	C
 	tm_creation_graph.dll!run_node_at_index(tm_creation_graph_instance_t * rc, unsigned int idx) Line 424	C
 	tm_creation_graph.dll!trigger_event(tm_creation_graph_instance_t * rc, unsigned __int64 e) Line 461	C
 	tm_creation_graph.dll!trigger_graph_event(tm_creation_graph_instance_t * inst, tm_creation_graph_context_t * context, unsigned __int64 event) Line 2695	C
 	tm_creation_graph.dll!refresh_outputs(tm_creation_graph_instance_t * inst, tm_creation_graph_context_t * context) Line 5027	C
 	tm_creation_graph.dll!output(tm_creation_graph_instance_t * inst, unsigned __int64 node_name_hash, tm_creation_graph_context_t * context, const tm_creation_graph_output_node_type_t * * output_node_type) Line 5058	C
 	tm_render_utilities.dll!private__refresh_draw_calls(tm_component_manager_o * manager, tm_render_component_with_graph_data_t * c, tm_creation_graph_context_t * ctx) Line 412	C
 	tm_render_utilities.dll!private__refresh_graph_output(tm_component_manager_o * manager, tm_render_component_with_graph_data_t * c, tm_creation_graph_context_t * ctx) Line 470	C
 	tm_render_utilities.dll!ci__fill_bounding_volume_buffer(tm_component_manager_o * manager, tm_render_args_t * args, const tm_entity_t * entities, const tm_transform_component_t * entity_transforms, const unsigned int * entity_indices, void * * render_component_data, unsigned int num_renderables, unsigned char * bv_buffer) Line 1300	C
 	tm_the_machinery_shared.dll!render_viewer_job(void * param) Line 975	C
```

I have a hard time finding some place in the callstack where it has already been set, so if I dig deeper I will start logging all lock and unlock callstacks for this type of atomic.

Will pause this attempt for now.

**Comment by niklas-ourmachinery** [comment]
@tobias-ourmachinery Can you investigate this and see if you get this too...
